src/common/buffer.cc

```diff
@@ -399,183 +458,380 @@ static simple_spinlock_t buffer_debug_lock = SIMPLE_SPINLOCK_INITIALIZER;
-	    << buffer::get_total_alloc() << bendl;
     }
+  }
+/*
+ * TODO - make crc32c calculation using kcrypto crc32c-intel algorithm, if possible
+ */
+  uint32_t crc32c(unsigned int off, unsigned int len,uint32_t init_crc) {
+    uint32_t crc;
+    get_data(); //make 'data' meaningfull
+    crc = ceph_crc32c(init_crc, reinterpret_cast<unsigned char*>(data) + off, len);
+    set_crc(make_pair(off, off + len), make_pair(init_crc, crc));
+    return crc;
+  }
 
-    ~raw_pipe() {
-      if (data)
-	free(data);
-      close_pipe(pipefds);
-      dec_total_alloc(len);
-      bdout << "raw_pipe " << this << " free " << (void *)data << " "
-	    << buffer::get_total_alloc() << bendl;
+  void append_pipe() {
+    int r;
+    if (pipe_curr_write_fd != -1) {
+      close(pipe_curr_write_fd);
+    }
+    int fds[2];
+    if (::pipe(fds) == -1) {
+      r = -errno;
+      bdout << "raw_pipe: error creating pipe: " << cpp_strerror(r) << bendl;
+      throw error_code(r);
+    }
+    r = ::fcntl(fds[write_end], F_SETFL, O_NONBLOCK);
+    if (r != -1)
+      r = ::fcntl(fds[read_end], F_SETFL, O_NONBLOCK);
+    if (r < 0) {
+      bdout << "raw_pipe: error setting nonblocking flag on temp pipe: "
+          << cpp_strerror(r) << bendl;
+      throw error_code(r);
     }
+    pipe_curr_write_fd = fds[write_end];
+    set_pipe_size(pipe_curr_write_fd);
+    pipe_read_elements.push_back({0, fds[read_end]});
+  }
 
-    bool can_zero_copy() const {
-      return true;
+
+  bool can_zero_copy() const {
+    return data == nullptr;
+  }
+
+  ssize_t consume_pipe(int fd, size_t len)
+  {
+    int devnull;
+    devnull = open("/dev/null", O_WRONLY);
+    if (devnull == -1)
+      return -1;
+    ssize_t r;
+    r = splice(fd, nullptr, devnull, nullptr, len, SPLICE_F_MOVE);
+    ::close(devnull);
+    return r;
+  }
+
+  bool is_pipe_full(int fd)
+  {
+    struct pollfd fds;
+    fds.fd = fd;
+    fds.events = POLLOUT;
+    fds.revents = 0;
+    int r = poll(&fds, 1, 0);
+    //when no fds.revents are set, means we are blocked
+    return (r == 0);
+  }
+
+  /*
+   * Streams data from buffer to specified file.
+   * If destination file is blocking, function will block until all data is transferred.
+   * If destination file is non-blocking, function will transfer as much data as possible.
+   *
+   * \note Transferred data is not lost from buffer.
+   *
+   * \param fd file to write to
+   * \param src_pos start streaming from this byte
+   * \param count amount of bytes to write
+   * \param dst_offset offset in \ref fd to write to. use -1 to append to current file position.
+   *
+   * \return amount of bytes transferred, or -1 if operation would block
+   *
+   * \throws end_of_buffer if request is outside buffer's range
+   */
+  ssize_t copy_to_fd(int fd, off_t ofs_from, ssize_t req_size, off_t dst_offset) {
+    if (ofs_from + req_size > len) {
+      throw end_of_buffer();
     }
+    if (data != nullptr) {
```

Comment:
There is difference between 'can_zero_copy()' and 'data!=nullptr'.

The first one is interface declaration, which reports whether this buffer::raw is somehow boosted. Determination when it is beneficial to report being 'can_zero_copy()' is subject to change.
The second one data != nullptr is querying internals of buffer::raw to determine where actual data resides.

The fact that can_zero_copy() is textually equal do data != nullptr in my opinion is insufficient to replace one with another.

I would to leave it as it is.